source/common/router/router.cc

```diff
@@ -446,7 +498,8 @@ void Filter::onRequestComplete() {
 }
 
 void Filter::onDestroy() {
-  if (upstream_request_) {
+  if (upstream_request_ &&
+      (!upstream_request_->stream_info_.lastUpstreamRxByteReceived() || !downstream_end_stream_)) {
```

Comment:
What's the case where this needs to be changed? This seems a little scary.

Reply:
Ah yeah, maybe you have a better suggestion of how to handle this.

unless we have a 2 phase canRecreateStream() -> recreateStream() we can't technically tell if the redirect is going to be handled, so we can't do
onUpstreamComplete()  // nulls out upstream_request_
callbacks->recreateStream() 
without screwing up the failure path because if recreateStream fails and we fall through to normal 302 handling, there's a bunch of code which assumes upstream_request_ is non-null.  if recreateStream succeeds, it destroys the filter "inline" and we don't get a chance to call onUpstreamComplete() before we hit onDestroy.

Reply:
Sorry I was doing a "quick pass" and thinking about this doesn't fall under the "quick pass" category. :) I will get back to you in a few days which I think is OK since we are punting for 1.9.0 anyway.

Reply:
OK I think I understand, basically, you are checking if we really need a reset, because otherwise upstream_request_ would already be null? I think this is correct, but it seems scary to me TBH. At the very least, can we add a bunch of comments around what is happening here? Next up, maybe an ASSERT in here that if the inverse of `!upstream_request_->stream_info_.lastUpstreamRxByteReceived() || !downstream_end_stream_)` is true, upstream_request is already nullptr? Another thought, any way to guard this with some bool that is tied to the redirect code directly such that it would be more clear?

Reply:
SGTM.  Overly verbose boolean and very detailed comment uploaded :-)